pkg/workflow/workflow.go

```diff
@@ -0,0 +1,166 @@
+	oamutil "github.com/oam-dev/kubevela/pkg/oam/util"
+	"github.com/oam-dev/kubevela/pkg/utils/apply"
+)
+
+type workflow struct {
+	app        *oamcore.Application
+	applicator apply.Applicator
+}
+
+// NewWorkflow returns a Workflow implementation.
+func NewWorkflow(app *oamcore.Application, applicator apply.Applicator) Workflow {
+	return &workflow{
+		app:        app,
+		applicator: applicator,
+	}
+}
+
+func (w *workflow) ExecuteSteps(ctx context.Context, rev string, objects []*unstructured.Unstructured) (bool, error) {
+	steps := w.app.Spec.Workflow
+
+	if len(steps) == 0 {
+		return true, nil
+	}
+
+	w.app.Status.Phase = common.ApplicationRunningWorkflow
+
+	w.app.Status.Workflow = []common.WorkflowStepStatus{}
+	for i, step := range steps {
+		obj := objects[i].DeepCopy()
+		obj.SetName(step.Name)
+		obj.SetNamespace(w.app.Namespace)
+		obj.SetOwnerReferences([]metav1.OwnerReference{{
+			APIVersion: oamcore.SchemeGroupVersion.String(),
+			Kind:       oamcore.ApplicationKind,
+			Name:       w.app.Name,
+			UID:        w.app.UID,
+			Controller: pointer.BoolPtr(false),
+		}})
+		err := w.applyWorkflowStep(ctx, obj, &types.WorkflowContext{
+			AppName:       w.app.Name,
+			AppRevision:   rev,
+			WorkflowIndex: i,
+		})
+		if err != nil {
+			return false, err
+		}
+
+		status, err := w.syncWorkflowStatus(step, obj)
+		if err != nil {
+			return false, err
+		}
+
+		w.app.Status.Workflow = append(w.app.Status.Workflow, *status)
+		switch status.Phase {
+		case common.WorkflowStepPhaseSucceeded: // This one is done. Continue
+		case common.WorkflowStepPhaseRunning: // Need to retry shortly.
+			return false, nil
+		default:
+			return true, nil
+		}
+	}
+	return true, nil // all steps done
+}
+
+func (w *workflow) applyWorkflowStep(ctx context.Context, obj *unstructured.Unstructured, wctx *types.WorkflowContext) error {
+	if err := addWorkflowContextToAnnotation(obj, wctx); err != nil {
+		return err
+	}
+
+	return w.applicator.Apply(ctx, obj)
+}
+
+func addWorkflowContextToAnnotation(obj *unstructured.Unstructured, wc *types.WorkflowContext) error {
+	b, err := json.Marshal(wc)
+	if err != nil {
+		return err
+	}
+	m := map[string]string{
+		oam.AnnotationWorkflowContext: string(b),
+	}
+	obj.SetAnnotations(oamutil.MergeMapOverrideWithDst(m, obj.GetAnnotations()))
+	return nil
+}
+
+func (w *workflow) syncWorkflowStatus(step oamcore.WorkflowStep, obj *unstructured.Unstructured) (*common.WorkflowStepStatus, error) {
+	status := &common.WorkflowStepStatus{
+		Name: step.Name,
+		Type: step.Type,
+		ResourceRef: runtimev1alpha1.TypedReference{
+			APIVersion: obj.GetAPIVersion(),
+			Kind:       obj.GetKind(),
+			Name:       obj.GetName(),
+			UID:        obj.GetUID(),
+		},
+	}
+
+	cond, found, err := utils.GetUnstructuredObjectStatusCondition(obj, "workflow-finish")
+	if err != nil {
+		return nil, err
+	}
+
+	if !found || cond.Status != "True" {
+		status.Phase = common.WorkflowStepPhaseRunning
+		return status, nil
+	}
```

Comment:
Hi, is there a way to skip this validation if the resource created in step has no status?

Reply:
No. We need to enforce this because the the workflow step needs some feedback on whether it finishes or not.

Reply:
Thanks for your reply, but then resources such as ConfigMap, TektonTask etc. will not be available in the workflow, and they will always running.

Is it possible to add a field to the WorkflowStepDefinition for skipping the validation of these resources?

Reply:
For such scenarios they are not put into workflow. They should be put into Component or Policy.

Reply:
I see, I will rethink the usage scenario of workflow, thanks!